src/compiler/scanner.ts

```diff
@@ -2373,15 +2428,965 @@ export function createScanner(languageVersion: ScriptTarget, skipTrivia: boolean
+
+            // DecimalEscape ::= [1-9] [0-9]*
+            function scanDecimalEscape(): boolean {
+                Debug.assertEqual(text.charCodeAt(pos - 1), CharacterCodes.backslash);
+                const ch = text.charCodeAt(pos);
+                if (ch >= CharacterCodes._1 && ch <= CharacterCodes._9) {
+                    const start = pos;
+                    scanDigits();
+                    decimalEscapes.push({ pos: start, end: pos, value: +tokenValue });
+                    return true;
+                }
+                return false;
+            }
+
+            // CharacterEscape ::=
+            //     | `c` ControlLetter
+            //     | IdentityEscape
+            //     | (Other sequences handled by `scanEscapeSequence`)
+            // IdentityEscape ::=
+            //     | '^' | '$' | '/' | '\' | '.' | '*' | '+' | '?' | '(' | ')' | '[' | ']' | '{' | '}' | '|'
+            //     | [~UnicodeMode] (any other non-identifier characters)
+            function scanCharacterEscape(): string {
+                Debug.assertEqual(text.charCodeAt(pos - 1), CharacterCodes.backslash);
+                let ch = text.charCodeAt(pos);
+                switch (ch) {
+                    case CharacterCodes.c:
+                        pos++;
+                        ch = text.charCodeAt(pos);
+                        if (isASCIILetter(ch)) {
+                            pos++;
+                            return String.fromCharCode(ch & 0x1f);
+                        }
+                        if (unicodeMode) {
+                            error(Diagnostics.c_must_be_followed_by_an_ASCII_letter, pos - 2, 2);
+                        }
+                        return String.fromCharCode(ch);
+                    case CharacterCodes.caret:
+                    case CharacterCodes.$:
+                    case CharacterCodes.slash:
+                    case CharacterCodes.backslash:
+                    case CharacterCodes.dot:
+                    case CharacterCodes.asterisk:
+                    case CharacterCodes.plus:
+                    case CharacterCodes.question:
+                    case CharacterCodes.openParen:
+                    case CharacterCodes.closeParen:
+                    case CharacterCodes.openBracket:
+                    case CharacterCodes.closeBracket:
+                    case CharacterCodes.openBrace:
+                    case CharacterCodes.closeBrace:
+                    case CharacterCodes.bar:
+                        pos++;
+                        return String.fromCharCode(ch);
+                    default:
+                        if (pos >= end || isLineBreak(ch)) {
+                            error(Diagnostics.Undetermined_character_escape, pos - 1, 1, ch);
+                            return "\\";
+                        }
+                        pos--;
+                        return scanEscapeSequence(/*shouldEmitInvalidEscapeError*/ unicodeMode, /*isRegularExpression*/ true);

```

Comment:
I don't know if we should always emit those invalid escape errors despite of flags (though that's actually what were intended originally), since it seems like a bad practice to use things like octal escape sequences in character class (currently those outside character class are not fell back from capturing group references to escape sequences even in non-Unicode mode either).